server/src/main/java/org/opensearch/common/cache/stats/StatsHolder.java

```diff
@@ -0,0 +1,200 @@
+    public void decrementSizeInBytes(List<String> dimensionValues, long amountBytes) {
+        internalIncrement(dimensionValues, (counter) -> counter.sizeInBytes.dec(amountBytes), false);
+    }
+
+    public void incrementEntries(List<String> dimensionValues) {
+        internalIncrement(dimensionValues, (counter) -> counter.entries.inc(), true);
+    }
+
+    public void decrementEntries(List<String> dimensionValues) {
+        internalIncrement(dimensionValues, (counter) -> counter.entries.dec(), false);
+    }
+
+    /**
+     * Reset number of entries and memory size when all keys leave the cache, but don't reset hit/miss/eviction numbers.
+     * This is in line with the behavior of the existing API when caches are cleared.
+     */
+    public void reset() {
+        resetHelper(statsRoot);
+    }
+
+    private void resetHelper(DimensionNode current) {
+        CacheStatsCounter counter = current.getStats();
+        counter.sizeInBytes.dec(counter.getSizeInBytes());
+        counter.entries.dec(counter.getEntries());
+        if (current.hasChildren()) {
+            // not a leaf node
+            for (DimensionNode child : current.children.values()) {
+                resetHelper(child);
+            }
+        }
+    }
+
+    public long count() {
+        // Include this here so caches don't have to create an entire CacheStats object to run count().
+        return statsRoot.getStats().getEntries();
+    }
+
+    private void internalIncrement(List<String> dimensionValues, Consumer<CacheStatsCounter> adder, boolean createNodesIfAbsent) {
+        assert dimensionValues.size() == dimensionNames.size();
+        boolean didIncrement = internalIncrementHelper(dimensionValues, statsRoot, 0, adder, createNodesIfAbsent);
+        if (didIncrement) {
+            adder.accept(statsRoot.getStats());
+        }
+    }
+
+    /**
+     * Use the incrementer function to increment/decrement a value in the stats for a set of dimensions.
+     * If createNodesIfAbsent is true, and there is no stats for this set of dimensions, create one.
+     * Returns true if the increment was applied, false if not.
+     */
+    private boolean internalIncrementHelper(
+        List<String> dimensionValues,
+        DimensionNode node,
+        int dimensionValuesIndex, // Pass in the relevant dimension index to avoid having to slice the list for each node.
+        Consumer<CacheStatsCounter> adder,
+        boolean createNodesIfAbsent
+    ) {
+        if (dimensionValuesIndex == dimensionValues.size()) {
+            return true;
+        }
+        DimensionNode child = node.getOrCreateChild(dimensionValues.get(dimensionValuesIndex), createNodesIfAbsent);
+        if (child == null) {
+            return false;
+        }
+        if (internalIncrementHelper(dimensionValues, child, dimensionValuesIndex + 1, adder, createNodesIfAbsent)) {
+            adder.accept(child.getStats());
```

Comment:
This should apply on the node itself, right?

Reply:
Sure we can change it to do this.